### PR TITLE
Catch malformed etcd_authority variables and raise a datastore error

### LINKED ISSUE #85: libcalico yields misleading error from bad ETCD_AUTHORITY env var
If `ETCD_AUTHORITY` environment is missing a colon, pycalico will fail to initialize the `IPAMClient` with  a misleading stack trace. 

```
$ ETCD_AUTHORITY=erroneous python -c "import pycalico.ipam; pycalico.ipam.IPAMClient()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/pycalico/datastore.py", line 137, in __init__
    (host, port) = etcd_authority.split(":", 1)
ValueError: need more than 1 value to unpack
```

[Link to line 137](https://github.com/projectcalico/libcalico/blob/master/calico_containers/pycalico/datastore.py#L137)

I propose that we should catch the value error and reword it to state "malformed ETCD_AUTHORITY", possibly in the form of a new MalformedEtcdAuthority exception?

### PR DESCRIPTION
**Pull request by ozdanborne**
Fixes #85

### CODE REVIEW

**Comment by TrimBiggs** [comment]
LGTM

**Comment by TrimBiggs** [comment]
We should have a test for this.

File: calico_containers/pycalico/datastore.py [hunk 1]
```diff
@@ -134,7 +134,15 @@ class DatastoreClient(object):
 
     def __init__(self):
         etcd_authority = os.getenv(ETCD_AUTHORITY_ENV, ETCD_AUTHORITY_DEFAULT)
-        (host, port) = etcd_authority.split(":", 1)
+        try:
+            (host, port) = etcd_authority.split(":", 1)
```
Comment (TrimBiggs) [inline calico_containers/pycalico/datastore.py]: I don't think this will catch if the authority is ":1234", but I think it will be pretty obvious when it fails elsewhere..

**Comment by TrimBiggs** [comment]
LGTM

File: calico_containers/pycalico/datastore.py (hunk 1 above)
Comment (caseydavenport) [inline calico_containers/pycalico/datastore.py]: I think the most robust way to do this is to define a regex pattern which matches valid etcd authorities and parse it using that.

For example, this won't catch if someone puts in `http://host:port` as their etcd authorit (which is a common and easy mistake for users to make)

**Comment by caseydavenport** [comment]
As I said in a comment, I think this should be using regex to do more comprehensive checking on the provided value.

File: calico_containers/pycalico/datastore.py (hunk 1 above)
Comment (caseydavenport) [inline calico_containers/pycalico/datastore.py]: Rather, it will likely catch the error below but the error message won't be great.

e.g

```
ERROR: Invalid ETCD_AUTHORITY port: 'host:port'
```

Rather than:

```
ERROR: ETCD_AUTHORITY 'http://host:port' should not include scheme 'http'
```

Alternatively, we might just want to handle that case gracefully...

Comment (TrimBiggs) [inline calico_containers/pycalico/datastore.py]: Good point.  Just need to keep in mind the valid characters for a hostname if using hostname vs ip.

**Comment by ozdanborne** [comment]
Maybe. We do, however, accept both hostnames and IPs. Therefore, our regex statement would  effectively be searching for two sets of values split by a colon, which is the same as what this code does.

**Comment by TrimBiggs** [comment]
There could be a utils `validate_etcd_authority` function:
Some regex check like `((\d{1,3}\.){3}(\d{1,3})|([\s\d]{8,32}):\d{1,5})` that checks both IP and hostname (hostname check there could be better or made need to be different).

This would be useful for when we allow a list of etcd authorities also.  We might want check for an optional `(http://|https://|etcd://){0,1}` if we start allowing this, but we could have wrappers around this function that get the IP/port regardless of what style of etcd authority is passed in.

**Comment by ozdanborne** [comment]
I realized that we already do `ETCD_AUTHORITY` validations in calico-containers utils.py, so I am changing this PR to copy those validations into libcalico for use by other orchestrators (in this case, libcalico itself). 

I'm going to try to keep the scope of this small so it is just a 1:1 move - the only change in this PR besides the move is calling the function from datastore.py ([see second commit](https://github.com/djosborne/libcalico/commit/8a01ecfaf0a6c619fcc99a27605b1e064c4a8650))

@TrimBiggs @caseydavenport I believe this takes care of all your comments (except multiple etcd_authorities and http/s, but I'd prefer to save those for a later PR). Sound good?

Additionally, I will open up a seperate PR to remove this functionality from calico_containers once this one is merged

File: calico_containers/pycalico/datastore.py [hunk 2]
```diff
@@ -134,7 +134,12 @@ class DatastoreClient(object):
 
     def __init__(self):
         etcd_authority = os.getenv(ETCD_AUTHORITY_ENV, ETCD_AUTHORITY_DEFAULT)
-        (host, port) = etcd_authority.split(":", 1)
+        if not validate_hostname_port(etcd_authority):
+            raise DataStoreError("Invalid %s. It must take the form <address>:<port>. "
+                            "Value provided is '%s'" % (ETCD_AUTHORITY_ENV,
```
Comment (TrimBiggs) [inline calico_containers/pycalico/datastore.py]: Alignment is off here

**Comment by TrimBiggs** [comment]
I'm ok with that.

This LGTM with a minor comment.  Are you planning to have a PR that removes this from calico-containers?

File: calico_containers/pycalico/util.py [hunk 3]
```diff
@@ -150,3 +150,44 @@ def validate_icmp_type(icmp_type):
...
+def validate_hostname_port(hostname_port):
+    """
+    Validate the hostname and port format.  (<HOSTNAME>:<PORT>)
+    An IPv4 address is a valid hostname.
+
+    :return: Boolean: True if valid, False if invalid
+    """
+    # Should contain a single ":" separating hostname and port
+    hostname_port = hostname_port.split(":")
+    if len(hostname_port) != 2:
```
Comment (caseydavenport) [inline calico_containers/pycalico/util.py]: Might be nice to have some debug logging in this function so we can track where validation went sour.

Comment (ozdanborne) [inline calico_containers/pycalico/util.py]: Might be nice to have logging in libcalico :frowning: Afaik we don't currently, right?

File: calico_containers/pycalico/util.py [hunk 4]
```diff
@@ -150,3 +150,44 @@ def validate_icmp_type(icmp_type):
...
+
+def validate_hostname_port(hostname_port):
+    """
+    Validate the hostname and port format.  (<HOSTNAME>:<PORT>)
+    An IPv4 address is a valid hostname.
+
+    :return: Boolean: True if valid, False if invalid
+    """
+    # Should contain a single ":" separating hostname and port
+    hostname_port = hostname_port.split(":")
```
Comment (caseydavenport) [inline calico_containers/pycalico/util.py]: Any possibility of something that is not a string being passed here?  Do we need to assert that `hostname_port` is a string before we try to split it?

File: calico_containers/pycalico/util.py (hunk 3 above)
Comment (caseydavenport) [inline calico_containers/pycalico/util.py]: We do have logging in libcalico (though it appears we don't in util.py).

See netns.py, ipam.py, handle.py, block.py

It is up to the calling code to display those logs or not.

**Comment by caseydavenport** [comment]
Looks fine to me.

I'd like to see the corresponding PR in `calico-containers` which removes this function.

File: calico_containers/pycalico/util.py (hunk 3 above)
Comment (ozdanborne) [inline calico_containers/pycalico/util.py]: I wouldn't mind adding logging to it. This PR or a new one?

**Comment by ozdanborne** [comment]
@caseydavenport I haven't opened the PR as it will just fail CI tests: https://github.com/projectcalico/calico-containers/compare/master...djosborne:remove-validate-hostnames?expand=1

File: calico_containers/pycalico/datastore.py [hunk 5]
```diff
@@ -134,7 +134,12 @@ class DatastoreClient(object):
 
     def __init__(self):
         etcd_authority = os.getenv(ETCD_AUTHORITY_ENV, ETCD_AUTHORITY_DEFAULT)
-        (host, port) = etcd_authority.split(":", 1)
+        if not validate_hostname_port(etcd_authority):
+            raise DataStoreError("Invalid %s. It must take the form "
+                                 "<address>:<port>. Value provided is '%s'" %
+                                 (ETCD_AUTHORITY_ENV, etcd_authority))
+        else:
```
Comment (caseydavenport) [inline calico_containers/pycalico/datastore.py]: You don't need the `else` here - happy to let it go for now but is cleaner to not include it.

File: calico_containers/pycalico/util.py [hunk 6]
```diff
@@ -150,3 +154,61 @@ def validate_icmp_type(icmp_type):
...
+    Validate the hostname and port format.  (<HOSTNAME>:<PORT>)
+    An IPv4 address is a valid hostname.
+
+    :return: Boolean: True if valid, False if invalid
+    """
+    # Should contain a single ":" separating hostname and port
+    if not isinstance(hostname_port, str):
+        _log.error("Must provide string for hostname:port validation, not: %s" % type(hostname_port))
+        return False
+    try:
```
Comment (caseydavenport) [inline calico_containers/pycalico/util.py:170]: Should have a newline before this try.

**Comment by caseydavenport** [comment]
Two minor comments - fix them if it pleases you.

Then, squash commits, and then LGTM.

File: calico_containers/pycalico/util.py (hunk 6 above)
Comment (ozdanborne) [inline calico_containers/pycalico/util.py:170]: Aesthetically, I'd agree that it looks better, and I'll change it. I'm curious if there's a pep8 that also says there should be a new line here? And what's the gist of it? Always add a new line after conclusion of an if statement?
